### DISCUSSION TITLE
Sector numbers starting with 193

### DISCUSSION THREAD

**Discussion by RCPDesigns** [keirf/greaseweazle#216] [Q&A] [answered]
I'm trying to read some old floppies from a Z80 CPM/2 system.  It uses a PD765 floppy disk controller setup for MFM Mode.
The floppy is double sided and 96TPI.  The manual mentions 512 bytes per sector and loosely implies that there are 15 sectors per track.   The PD765 data sheet implies there are 8 sectors per track. The manual says the disk holds 800K. 

I've gone through all the formats and the ones that "find something" all have the messages...

T0.0: Ignoring unexpected sector C:0 H:0 R:193 N:2
.
.
.
T0.0: Ignoring unexpected sector C:0 H:0 R:202 N:2

I assume R is for sector. The lowest sector number is 193 and the highest is 202.  All sector numbers between 193 and 202 are found, usually twice. Assuming N is the bytes/sector code... that looks correct as the PD765 data sheet says 02 is for 512 bytes/sector.

My question is what would explain the sector numbers being so high?

I've looked at the source and it is very well laid out so I think that I can tweak the MFM parameters if given some direction as to why I'm seeing the high sector numbers.

**Comment by keirf** [comment, marked as answer]
Although most systems start sector numbering from 0 or 1, it isn't universal. There is no requirement for sector numbering to start from a particular base, nor even to be consecutive. On this system the numbering starts at 193d = C1h. It's not a problem. Amstrad CPC does this too iirc.

**Reply by RCPDesigns** [reply]
Should I add a `startingSector` variable such that the sectors will not be ignored?

**Reply by keirf** [reply]
The relevant field in the class definitions in `greaseweazle/codec/ibm/mfm.py` is `id0`.

**Reply by RCPDesigns** [reply]
Perfect!  I created a new format and it is working!  It missed 2 sectors on the first disk I tried.  I'll try some more to see if I might need to tweak the gaps and/or speed some.
It is a pleasure to work with your code.  I appreciate the thought you put into it so that it could be easily extendable and worked on by others.
Once I have the format tweaked would you like me to create a branch and PR the format back to `master`?

**Reply by RCPDesigns** [reply]
It seems I have inadvertently run into the format for MSDOS 5.0 disk.  It is just like PC.98 but `id0=1` `nsec=15` and `sz=2`
A few of the disk I have in this box are for an old PC system and not the DataVue 80 as I thought.

I've processed a dozen disks and while some have zero issues it seems several have issues on tracks 2-4.  I guess it could be wear in that area.  Are there some parameters (gap, skew, etc) that might affect the first few tracks more?

**Reply by keirf** [reply]
You can dump to SCP, load into HxC tools, and then use the track analyser to view the track and look for wobble or corruption in the "bit bands" in the flux visualisation.

Do you have the disk rpm set correctly? 300rpm vs 360rpm?

**Reply by RCPDesigns** [reply]
I have `--adjust-speed 300`   The original drive was 300 rpm and the drive I'm reading with is 360 rpm. 

I loaded the scp into HxC and it looked pretty bad. 

![image](https://user-images.githubusercontent.com/4529546/185439658-0f52dca0-2350-4f97-931c-424cacbe1719.png)

I reduced `gap3` from 116 to 42 and `gw` reported more missed sectors (15 out of 1600) but when loaded in HxC it said it was perfect??

![image](https://user-images.githubusercontent.com/4529546/185439932-1c5b1bfd-dc57-4c6c-af07-f64033d4694e.png)

That's great (I guess) but what the heck to I do now?  lol

**Reply by keirf** [reply]
15 sectors is weird choice for 300rpm. Or are we talking about 10-sector disks again?

If you `gw read my.scp` with no `--format=` option then the SCP is raw and none of the ibm mfm codec in greaseweazle comes into play at all. You can still `--adjust-speed` if you want.

Which version of greaseweazle tools are you using?

**Reply by keirf** [reply]
I can also guarantee a gap3 of 116 is wrong for 512-byte sectors. That is usual for 1024-byte sectors. 30-40 bytes ish is more like it for 10-sector 300rpm tracks.

**Reply by RCPDesigns** [reply]
Yes, we're back to the 10 sector disk.  I checked out `master` this morning.  I updated the hardware with `gw update` last night.
I can't seem to convince MacOS to let HxC export a file.   So that seemingly perfect image up there only exists in HxC.
-edit- looks like HxC is failing to export the Atari ST format.  The raw `.img` format works and I guess works for me.
My format looks like this right now.
```class DV80_800(IBM_MFM_720):

    clock = 2e-6
    time_per_rev = 0.2
    gap_3 = 42
    nsec  = 10
    id0   = 193
    sz    = 2

**Reply by keirf** [reply]
I'd suggest `gap_3 = 30`. But for checking in HxC I would suggest raw dump, no `--format=` option which would make `gap_3` unused anyway.

If you can get a raw dump `gw read my.scp` perhaps just with a `--tracks=` parameter, and upload it, I can take a look.

**Reply by keirf** [reply]
Because if you specify `--format=` then your SCP is "cooked"/regenerated, and not actually a raw dump. Unless you also specify `--raw`.

**Reply by keirf** [reply]
You can specify `--adjust-speed=` alongside `--raw` though. That's legit, if your dump drive is wrong speed.

**Reply by RCPDesigns** [reply]
ok, doing that now.  Couldn't figure out `--tracks` so there will be a couple extra tracks. 
Thanks for taking the time to look at it.  Where do I upload?

**Comment by RCPDesigns** [comment]
[raw3.scp.zip](https://github.com/keirf/greaseweazle/files/9375657/raw3.scp.zip)

disk format is only 80 tracks so 2 extra tracks in the read

**Reply by keirf** [reply]
Load this into HxC tools, and look at the disk view. Note the red bits. Switch to stream view and look at some tracks on side 1. Look at how on bad (red) sectors the bit bands "drop out" and flux turns to noise. Could be due to mould on disk. Sometimes cleaning the disk surface can fix this.

The good HxC visualisations were probably because Greaseweazle had regenerated the tracks. However it would have written filler bytes into sectors it could not decode in the raw dump. So it was a false sense of success.

**Reply by RCPDesigns** [reply]
Great, I'll try cleaning the disk.  Thanks a ton for the help!
GW is one of the coolest things I've bought in a long time.  Last week I managed to read in 50 disks from 1980 for my first computer, the Intecolor 3650.  The atari.90 format worked perfectly.  I load them into my emulator and was back in high-school immediately.  Some of those disks were marked "bad" so the flux technology is amazing!

**Reply by RCPDesigns** [reply]
Cotton ball and some isopropyl alcohol and a perfect read.  Only one retry.  The fun thing about the disk image is you can see where to look on the physical disk. 

![image](https://user-images.githubusercontent.com/4529546/185465988-2fd1b8ec-d1f2-4a70-b1c1-5e86866e00cc.png)
